client/serviceregistration/checks/client_test.go

```diff
@@ -327,16 +327,26 @@ func TestChecker_Do_TCP(t *testing.T) {
 	}
 }
 
-func tcpServer(ctx context.Context, port int) {
+// tcpServer will start a tcp listener that accepts connections and closes them.
+// The caller can close the listener by cancelling ctx.
+func tcpServer(t *testing.T, ctx context.Context, port int) {
 	var lc net.ListenConfig
-	l, _ := lc.Listen(ctx, "tcp", net.JoinHostPort(
+	l, err := lc.Listen(ctx, "tcp", net.JoinHostPort(
 		"localhost", fmt.Sprintf("%d", port),
 	))
-	defer func() {
+	must.NoError(t, err, must.Sprint("port", port))
+	t.Cleanup(func() {
 		_ = l.Close()
-	}()
-	con, _ := l.Accept()
-	defer func() {
-		_ = con.Close()
+	})
+
+	go func() {
+		// caller can stop us by cancelling ctx
+		for {
+			con, conErr := l.Accept()
+			if conErr != nil {
+				return
+			}
+			_ = con.Close()
```

Comment:
We're closing this connection immediately after accepting. Will the code under test be ok with this if it gets the ACK and then an immediate RST? Alternatively, can we lean on the context to close out the listener and all the connections so that we know they'll stick around until we're done with the test?

Reply:
Oh good call, let's do that